Add match-contract

eval opened the pull request:
For a recent project this contract would have been handy:

```ruby
RGB = Match[/\#h{6}/] # e.g. '#ffffff'

Contract RGB => nil
def assign_color(color)
  # 
end
```

Other example:
```ruby
Contract Match[default_mail_regexp] => nil
def send_invitation(email)
...
```

As it's not easy to get the effect of a logical AND with regular expressions, so I added that as well:

```ruby
PartnerEmail = Match[default_mail_regexp, /(company1|company2)\.com\z/]

Contract PartnerEmail => nil
def invite_to_project(email)
```

Hopefully it's useful to others! If so I'm happy to move the explanation to the tutorial.

nixpulvis commented:
Could we use the bare regex, for example:

```ruby
Contract /\#h{6}/ => nil
```

waterlink commented:
According to:
https://github.com/egonSchiele/contracts.ruby/blob/master/lib/contracts/validators.rb#L3-L57
- it is not possible to match with bare regex currently. In case we find it
as an appropriate way to match regex, one could just add new key with
corresponding lamda there.

--
Best Regards,
Alexey Fedorov,
Sr Ruby, Clojure, Crystal, Golang Developer,
Microservices Backend Engineer,
+49 15757 486 476

2015-08-31 2:05 GMT+02:00 Nathan Lilienthal <notifications@github.com>:

> Could we use the bare regex, for example:
>
> Contract /\#h{6}/ => nil
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/egonSchiele/contracts.ruby/pull/196#issuecomment-136219602>
> .
>

egonSchiele commented:
I like this contract! It would be a lot more elegant as a bare regex as @nixpulvis suggested. You could still get the AND functionality by using the And contract: `And[/some regex/, /another regex/]`. Do you feel comfortable adding your contract to this hash, @eval ?

https://github.com/egonSchiele/contracts.ruby/blob/master/lib/contracts/validators.rb#L3-L57

eval commented:
Much cleaner indeed, will update. Thanks for the pointers!

eval commented:
I updated the code and added some explanation to `TUTORIAL.md` /cc @egonSchiele 

Note that the current version cannot be used in something like
```ruby
Contract (a: C::Match[/foo/]) => nil
```
Should we support this as well?

waterlink commented:
Hm, I think validators should work correctly inside of hashes. So this should work:

```ruby
Contract {a: /foo/} => nil
def bar(hash)
end
```

waterlink commented on spec/builtin_contracts_spec.rb:
```diff
@@ -360,6 +360,16 @@
     end
   end
 
+  describe "Regexp:" do
```
Can you move these specs to `validators_spec`?

waterlink commented:
Yeah, hash validator here treats values as contracts too, so it is possible to have them as a regexp and regexp validator will kick in for them. Here: https://github.com/eval/contracts.ruby/blob/feature/match-contract/lib/contracts/validators.rb#L19-L26

eval commented:
@waterlink this works indeed within other contracts :+1:

I moved the specs. Let me know if there's anything else.

waterlink commented on spec/validators_spec.rb:
```diff
@@ -1,9 +1,9 @@
-module Contracts
-  RSpec.describe "range contract validator" do
-    subject(:o) { GenericExample.new }
+RSpec.describe 'Contract validators' do
```
rubocop says here to switch to double quote string.

waterlink commented on spec/validators_spec.rb:
```diff
@@ -22,4 +22,26 @@ module Contracts
+
+    describe 'within a hash' do
+      it "should pass for a matching string" do
+        expect { o.hash_containing_foo(host: "foo.example.org") }.to_not raise_error
```
Can you use here `1.8` hash rocket syntax, since we still want to support `1.8` ?

eval commented:
Amended the style-fixes - thanks for your feedback @waterlink

waterlink commented:
LGTM

waterlink commented:
Thanks @eval Good work!
